enterprise/views.py

```diff
@@ -1137,3 +1138,136 @@ def post(self, request, enterprise_uuid, program_uuid):
+
+        We check for the following criteria:
+        - The `audit` query parameter.
+        - The user's being routed to the course enrollment landing page.
+        - The customer's catalog contains the course in question.
+        """
+        request_has_audit_query_param = request.GET.get('audit')
+        is_course_enrollment_url = request.path == self.COURSE_ENROLLMENT_VIEW_URL.format(
+            enterprise_customer.uuid,
+            course_id
+        )
+        customer_catalog_contains_course = enterprise_customer.catalog_contains_course(course_id)
+        return request_has_audit_query_param and is_course_enrollment_url and customer_catalog_contains_course
+
+    def enroll_learner(self, request, enterprise_customer, course_id, mode):
+        """
+        Enroll a user into a course track, and register enterprise course enrollment & data sharing consent objects.
+        """
+        with transaction.atomic():
+            enterprise_customer_user, __ = EnterpriseCustomerUser.objects.get_or_create(
+                enterprise_customer=enterprise_customer,
+                user_id=request.user.id
+            )
+
+        enrollment_api_client = EnrollmentApiClient()
+
+        # Check to see if the user's already enrolled and we have an enterprise course enrollment to track it.
+        enrolled_in_course = enrollment_api_client.is_enrolled(request.user.username, course_id)
+        enterprise_course_enrollment_exists = EnterpriseCourseEnrollment.objects.filter(
+            enterprise_customer_user__enterprise_customer=enterprise_customer,
+            enterprise_customer_user__user_id=request.user.id,
+            course_id=course_id
+        ).exists()
+        if not (enrolled_in_course and enterprise_course_enrollment_exists):
+            # Directly enroll into the audit track.
+            EnterpriseCourseEnrollment.objects.update_or_create(
+                enterprise_customer_user=enterprise_customer_user,
+                course_id=course_id,
+            )
+            enrollment_api_client.enroll_user_in_course(request.user.username, course_id, mode)
+            self.track_event(request.user.id, 'edx.bi.user.enterprise.enrollment.course', {
+                'category': 'enterprise',
+                'label': course_id,
+                'enterprise_customer_uuid': str(enterprise_customer.uuid),
+                'enterprise_customer_name': enterprise_customer.name,
+                'mode': mode,
+            })
+
+        # The courseware view logic will check for DSC requirements, and route to the DSC page if necessary.
+        return redirect(LMS_COURSEWARE_URL.format(course_id=course_id))
+
+    def redirect(self, request, *args, **kwargs):
+        """
+        Redirects to the appropriate view depending on where the user came from.
+        """
+        enterprise_customer_uuid, resource_id = RouterView.get_path_variables(**kwargs)
+        for path, view_class in self.VIEWS.items():
```

Comment:
It would be better if we didn't have to loop through all the views here. Could we use a regular expression to replace the `enterprise_customer_uuid` and `resource_id` values in the path with `{}`, then use the result as the key on the `VIEWS` dict?

Reply:
Yep, great idea.